components/feature/readerview/src/main/java/mozilla/components/feature/readerview/ReaderViewFeature.kt

```diff
@@ -0,0 +1,141 @@
+        }
+
+        var fontType by Delegates.observable(FontType.valueOf(prefs.getString(FONT_TYPE_KEY, "SANS_SERIF")!!)) {
+            _, old, new -> saveAndSendMessage(old, new, FONT_TYPE_KEY)
+        }
+
+        @Suppress("UNUSED_PARAMETER")
+        private fun saveAndSendMessage(old: Any, new: Any, key: String) {
+            if (old != new) {
+                // TODO save shared preference
+                // TODO send message to reader view web extension
+            }
+        }
+
+        companion object {
+            const val COLOR_SCHEME_KEY = "mozac-readerview-colorscheme"
+            const val FONT_TYPE_KEY = "mozac-readerview-fonttype"
+            const val FONT_SIZE_KEY = "mozac-readerview-fontsize"
+        }
+    }
+
+    override fun start() {
+        if (!ReaderViewFeature.installed) {
+            ReaderViewFeature.install(engine)
+        }
+        observeSelected()
+    }
+
+    override fun onBackPressed(): Boolean {
+        // TODO send message to exit reader view (-> see ReaderView.hide())
+        return true
+    }
+
+    override fun onSessionAdded(session: Session) {
+        // TODO make sure we start listening to reader view web extension
+        // messages for this session: Here we will just hook things up
+        // using the readerViewWebExt WebExtension object which will
+        // provide the functionality for bi-directional messaging
+        // (bridge to GeckoView's setSessionMessageDelegate).
+    }
+
+    override fun onSessionRemoved(session: Session) {
+        // TODO stop listening to reader view web extension messages of this session
+    }
+
+    fun showReaderView() {
+        // TODO send message to show reader view (-> see ReaderView.show())
+    }
+
+    fun hideReaderView() {
+        // TODO send message to hide reader view (-> see ReaderView.hide())
+    }
+
+    // TODO observe messages delivered from the web extension to the selected session
+    // e.g. to know whether or not reader view is available for the current page.
+    // The session will be updated by the messaging delegate integration logic in our
+    // WebExtension object. So here we can just observe the message on the session.
+    // override fun onWebExtensionMessage() { }
+
+    companion object {
+        @VisibleForTesting
+        internal val readerViewWebExt =
+                WebExtension("mozac-readerview", "resource://android/assets/extensions/readerview/")
+
+        @Volatile
+        internal var installed = false
+
+        /**
+         * Installs the readerview web extension in the provided engine.
+         *
+         * @param engine a reference to the application's browser engine.
+         */
+        fun install(engine: Engine) {
+            engine.installWebExtension(readerViewWebExt,
+                onSuccess = {
+                    Logger.debug("Installed extension: ${it.id}")
+                    installed = true
```

Comment:
Since this callback is asynchronous, it is still technically possible to call `installWebExtension` twice here..?

Reply:
OK, I've fixed this for now, but I think I will try to move the installation state into our `WebExtension` abstraction when we add more more functionality to it (as part of #2243)